graylog2-server/src/main/java/org/graylog2/inputs/codecs/JsonPathCodec.java

```diff
@@ -25,48 +28,68 @@
 import org.graylog2.plugin.configuration.Configuration;
 import org.graylog2.plugin.configuration.ConfigurationRequest;
 import org.graylog2.plugin.configuration.fields.ConfigurationField;
+import org.graylog2.plugin.configuration.fields.DropdownField;
 import org.graylog2.plugin.configuration.fields.TextField;
 import org.graylog2.plugin.inputs.annotations.Codec;
 import org.graylog2.plugin.inputs.annotations.ConfigClass;
 import org.graylog2.plugin.inputs.annotations.FactoryClass;
 import org.graylog2.plugin.inputs.codecs.AbstractCodec;
 import org.graylog2.plugin.inputs.codecs.CodecAggregator;
 import org.graylog2.plugin.journal.RawMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import javax.inject.Inject;
-import java.nio.charset.StandardCharsets;
-import java.util.List;
-import java.util.Map;
+import java.io.IOException;
+import java.util.*;
 
 @Codec(name = "jsonpath", displayName = "JSON Path")
 public class JsonPathCodec extends AbstractCodec {
 
+    private static final Logger LOG = LoggerFactory.getLogger(JsonPathCodec.class);
     public static final String CK_PATH = "path";
     public static final String CK_SOURCE = "source";
+    public static final String CK_MODE = "";
 
     private final JsonPath jsonPath;
+    private final String modeString;
+    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
 
     @AssistedInject
     public JsonPathCodec(@Assisted Configuration configuration) {
         super(configuration);
         final String pathString = configuration.getString(CK_PATH);
         jsonPath = pathString == null ? null : JsonPath.compile(pathString);
+        modeString = configuration.getString(CK_MODE);
     }
 
     @Nullable
     @Override
     public Message decode(@Nonnull RawMessage rawMessage) {
-        if (jsonPath == null) {
-            return null;
+        Map<String, Object> fields = new HashMap<>();
+        if (Objects.equals(modeString, "path")) {
+            if (jsonPath == null) {
+                return null;
+            }
+            final String json = new String(rawMessage.getPayload(), charset);
+            fields = read(json);
+        } else if (Objects.equals(modeString, "full")) {
+            final String json = new String(rawMessage.getPayload(), charset);
+            try {
+                fields = useFlattener(json);
+            } catch (JsonTypeException e) {
+                LOG.warn("JSON contains type not supported by JsonFlattenService.flatten.");
```

Comment:
You should always include the original exception (`e`) in the log message you are throwing, otherwise the user is missing context.